src/PowerShellEditorServices/Workspace/Workspace.cs

```diff
@@ -625,6 +625,64 @@ private static string UnescapeDriveColon(string fileUri)
             return sb.ToString();
         }
 
+        /// <summary>
+        /// Converts a file system path into a DocumentUri required by Language Server Protocol.
+        /// </summary>
+        /// <remarks>
+        /// When sending a document path to a LSP client, the path must be provided as a
+        /// DocumentUri in order to features like the Problems window or peek definition
+        /// to be able to open the specified file.
+        /// </remarks>
+        /// <param name="path">
+        /// A file system path. Note: if the path is already a DocumentUri, it will be returned unmodified.
+        /// </param>
+        /// <returns>The file system path encoded as a DocumentUri.</returns>
+        internal static string ConvertPathToDocumentUri(string path)
+        {
+            const string fileUriPrefix = "file:///";
+
+            if (path.StartsWith("untitled:", StringComparison.Ordinal))
+            {
+                return path;
+            }
+
+            if (path.StartsWith(fileUriPrefix, StringComparison.Ordinal))
+            {
+                return path;
+            }
+
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                // On a Linux filesystem, you can have multiple colons in a filename e.g. foo:bar:baz.txt
+                string absoluteUri = new Uri(path).AbsoluteUri;
+
+                // First colon is part of the protocol scheme, see if there are other colons in the path
+                int firstColonIndex = absoluteUri.IndexOf(':');
+                if (absoluteUri.IndexOf(':', firstColonIndex + 1) >= 0)
+                {
+                    absoluteUri =
+                        absoluteUri.Substring(0, firstColonIndex + 1) +
+                        absoluteUri.Substring(firstColonIndex + 1).Replace(":", "%3A");
```

Comment:
Would be nice to cut out some string allocations here

```csharp
absoluteUri = new StringBuilder()
    .Append(absoluteUri, firstColonIndex + 1, absoluteUri.Length - firstColonIndex)
    .Replace(":", "%3A")
    .Insert(0, absoluteUri, firstColonIndex + 1)
    .ToString();
```

Or maybe `Append` the whole `absoluteUri` and use the `Replace` overload that takes index/count

Reply:
Good suggestion!  I like it.

Reply:
Interesting that `Insert()` doesn't do what you (or I) think it does.  It inserts **copies** of the string where the third parameter is the number of copies.  So going to have one more alloc - `.ToCharArray()`.

Reply:
> Interesting that Insert() method doesn't do what you (or I) think it does. It inserts copies of the string where the third parameter is the number of copies.

Yeah, didn't expect that! Good catch 🙂